### ISSUE TITLE
Only show candidates in range for best candidate facet

### ISSUE THREAD

**Issue by gitonthescene** [OpenRefine/OpenRefine#3673]
After reconciling you're presented with a numerical facet called "best candidate's score" which you can use to adjust the range of which rows to show.  But currently it only limits which _rows_ to show and not which _candidates_ to show.

### Proposed solution

It would be nice if this facet limited _both_ which rows and which candidates to show so that only rows which have some candidates within this range (the current behavior) but also only show those candidates in this range as well.

### Additional context

This is mostly a "nice to have".  I think it reduces cognitive dissonance just a hair and also makes it a tiny bit easier to aim the mouse at the choice you want by removing clutter.

I'm guessing this is mostly a change on the client, though I'm not sure how accessible the range is when the rows are being rendered.

**Comment by gitonthescene** [comment]
If I'm reading the code right, this should be accessible from anywhere with [`ui.browsingEngine.getJSON()`](https://github.com/OpenRefine/OpenRefine/blob/82b1cf0bfaf3c8ab57a12e0f2004bdeb71432383/main/webapp/modules/core/scripts/project/browsing-engine.js#L157).

From the browser web console, it looks like it should have enough info in it as well.

``` sh
>> ui.browsingEngine.getJSON()
{
facets: (3) […]
0: {…}
columnName: "countyLabel"
expression: "cell.recon.best.score"
from: 81
name: "countyLabel: best candidate's score"
selectBlank: true
selectError: true
selectNonNumeric: true
selectNumeric: true
to: 101
type: "range"
<prototype>: Object { … }
1: Object { type: "list", name: "matchStateID", columnName: "matchStateID", … }
2: Object { type: "text", name: "XcountyName", columnName: "XcountyName", … }
length: 3
<prototype>: Array []
mode: "row-based"
<prototype>: Object { … }
>>
```

**Comment by wetneb** [comment]
Interesting! I am reluctant to this change, for various reasons.
* first, as a user it's not really clear to me that I would want this most of the time: when I review high-scoring candidates it is useful to also see low-scoring ones for the same row. The correct match might have a low score, and it might be more obvious visually that the low-scoring one is correct than that the high-scoring one is incorrect.
* second, this would go against the principle that facets only filter rows/records as a whole and do not tweak their contents when displaying them. I am not saying that this is written in stone anywhere but if we want to question that idea then we should probably step back a bit and design something that is more flexible (generic filtering capabilities for candidates? being able to reorder, rescore candidates? I think there are many exciting things to consider there!)
* third, this would need to be properly integrated in the backend to avoid being confusing. Say I select only the high scoring candidates and then derive a new column based on the list of candidates: there will be a discrepancy between the frontend (where I only see the high-scoring candidates) and the backend (where my GREL expression sees all of them).

**Comment by ostephens** [comment]
>second, this would go against the principle that facets only filter rows/records as a whole and do not tweak their contents when displaying them. I am not saying that this is written in stone anywhere but if we want to question that idea then we should probably step back a bit and design something that is more flexible (generic filtering capabilities for candidates? being able to reorder, rescore candidates? I think there are many exciting things to consider there!)

I think this is key - making this change completely changes the behaviour of the facet/filter functionality in this one case. This seems like a bad idea.

Linking this view to the facet also means that it will *only* be available in this context. Given the issue as described, it would make more sense to me to consider a View option which displays only the best candidate - or possibly even (or additionally) "all candidates scoring above X" type option for more flexibility

**Comment by gitonthescene** [comment]
I'm more than willing to discuss alternatives.  As a user though, it struck me as odd that the facet would limit the rows but still show me candidates outside of the restricted range.  As a programmer I can appreciate the trying to keep the functionality orthogonal but I don't think this is something users are thinking about.

Also I don't think the average user really thinks about whether the matching candidates are part of the row.  It comes across as more of a useful view of related data.

@ostephens - Can you think of any other examples where something like this _might_ apply?  It is possible that this facet is unique in this regard.  I'm not exactly sure what you mean by a "View option".  Are there examples of other View options I can compare with or are you proposing a new interface similar to what @wetneb is alluding to above?

It sounds to me like the suggestion is to basically have a "facets" on the listed candidates in addition to the rows.  If these two different facets are treated as orthogonal, I can imagine the vast majority of people wanting to keep them in sync, though I admit this may be just my bias.  Would you really want to see _rows_ which have _some_ candidate with a score above X but then show candidates with a score above (or below) Y?

Again, it might be my bias, but I'm assuming that people using the current facet are explicitly only interested in the candidates fitting in the range of the facet.

**Comment by gitonthescene** [comment]
It's probably worth being explicit that this is only a view issue, the candidates are not being changed at all.  Further the problem in @wetneb's third point above is pretty obviously addressable by using a `filter()` in your GREL expression and I don't have a strong feeling about this being a surprise for users.

**Comment by wetneb** [comment]
I totally agree that it feels natural to somehow couple the filtering done on rows and the one done inside the candidates. I think it's really exciting to start thinking about this actually! I think we should be able to find something that fits your bill and can also accommodate more general filtering needs.

**Comment by gitonthescene** [comment]
I'm happy to let this sit if we want to mull this over.  It's definitely a nice-to-have for me more than a killer feature.

**Comment by gitonthescene** [comment]
Thinking about this a bit more, I see that the filter is for "___best___ candidates score".  I.e. it only claims to be looking at the _best_ score among the candidates and says nothing about the other candidates.  My reading was more like "candidates score" and using the fact that if the "best candidate" was in range then there is at least one candidate in range.  Though this does mean that if your range doesn't include the best score for the row, then the row won't appear even if it does have *some* candidate in range.

Maybe it's better to have have a "candidates score" facet?

**Comment by wetneb** [comment]
That sounds much easier, as it should be doable to have a candidates score facet without changing anything to the architecture. A GREL expression that returns a list of numeric values should be treated by the facet accordingly.

**Comment by ostephens** [comment]
>@ostephens - Can you think of any other examples where something like this might apply? It is possible that this facet is unique in this regard

The most obvious possibility would be only showing the rows within a record that match the facet value/filter - although this is slightly different because in that case each value is in it's own cell (and so in this scenario you would be hiding/showing rows rather than values within cells). Equally I guess that any cell that contains (from the users perspective) multiple values could be something that the user would want to do - but it's a less obvious use case I think.

However - reflecting on it, obviously what we are doing here is filtering a list of possible matches, not filtering out content from the cells - and of course the reconciliation process has precedent for this (i.e. once you've selected a match, the other potential matches no longer display). So perhaps my original objection to this has being counter to existing behaviour is too strong.

>I'm not exactly sure what you mean by a "View option"

In the dropdown menu at the top of each column there is a "View" option. This currently controls hiding/showing columns.
In the "View" option in the "All" column drop down there is an additional option to hide/show null values in cells. 

>Would you really want to see rows which have some candidate with a score above X but then show candidates with a score above (or below) Y?

In short - Yes! I want to look at the rows that have a match in a certain score/confidence range, but that doesn't necessarily mean that's the correct match. Where I want to just use the best match then there is the existing "Match each cell to its best candidate" option.

As long as I'm in the position of making decisions about matches (rather than just relying on the score) then I want to see the range of options that has been found.

I still feel at the moment like a "View best candidates only" should be kept separate to "Filter rows to those where the best candidates score falls into a range". However, I do understand the argument for these being closely coupled. Maybe a checkbox on the filter that controls the view?

**Comment by gitonthescene** [comment]
> In short - Yes! I want to look at the rows that have a match in a certain score/confidence range, but that doesn't necessarily mean that's the correct match. Where I want to just use the best match then there is the existing "Match each cell to its best candidate" option.

I must admit it still sounds strange to me to not want to see rows where _all_ the candidates score low but to want to see low scoring candidates for rows which have _some_ candidate which scores well.  If you're conceding that low scoring candidates might be the right match, wouldn't that apply to all rows?  And if that's not your immediate concern, shouldn't seeing candidates in range suffice?

That said, I agree that my proposal doesn't quite fit the "best candidate" facet for the reason described in my last post.  The new proposal is to have essentially a "some candidate" facet where you see rows which have _some_ candidate in the facet range and to tie this facet with the view on the candidates to see only those candidates fitting in that range.

**Comment by wetneb** [comment]
> If you're conceding that low scoring candidates might be the right match, wouldn't that apply to all rows?

I often feel like I do not have enough time to review all rows - I am lazy and only want to spend time on rows for which there is a good chance of finding a match. But when I do decide to match a cell to a candidate I want to do it in confidence that I am not overseeing candidates for that row.

This is an attitude I have when I want high precision but do not care much about recall.

**Comment by gitonthescene** [comment]
@wetneb I guess that makes sense but I'm still confused by the mix of being lazy and disciplined.  It feels like if you find a match outside of the range you've restricted it to, it's basically luck.  I would trust my eyes for spot checks but if they're the only things assuring accuracy I don't hold out a lot of hope.  

So are the rows which don't have some candidate high enough left for dead though they're equally likely to have a match outside of the facet range?

I would generally do a first pass of "likely" candidates and then have some overall check for accuracy, assuming I had some heuristic.  Once you hit a certain size (and I might say that's the size at which you want to use filtering) you probably need a heuristic because eyeballing that many rows is difficult.

**Comment by ostephens** [comment]
I think if we can find a way of offering this as an option which can be used if the user wants to (even happy for it to be the default behaviour as long as it can be overridden) then I think it's a good idea.

I guess ultimately people have different preferences for working through options and dividing up their work. For me, dealing with lines with high scoring candidates is a way of isolating the rows which are likely to be the fastest to work through - but once I've isolated them I want to work through them all.

Overall hiding candidates that don't meet a (user-selected) score threshold seems like a good idea. So I think really this is just about how we implement this with the flexibility to allow the user to show the hidden candidates when they want. I still think that this should be an option that can be used independently of any particular facet - because it could generally be useful no matter whether I've applied a facet or not, but I'd be very happy to see it integrated with specific facets to let the user keep the selections in sync.

**Comment by wetneb** [comment]
> For me, dealing with lines with high scoring candidates is a way of isolating the rows which are likely to be the fastest to work through - but once I've isolated them I want to work through them all.

That's a much better way to phrase my comment above, thanks!

**Comment by gitonthescene** [comment]
@ostephens - I agree this should be a different facet and don’t mean to suggest that this replace the current one.  I was just trying to understand the suggest workflow for the current facet.
